### PR TITLE
Refactor vendor dir

### PR DESCRIPTION
**Pull request by mara214**
## Changes

This pr refactors the package store from:

```
proto
  /vendor/
    <a>/
      Proto.toml
      vendor/<c>/..
    <b>/
      Proto.toml
      vendor/<c>/..
```

To 

```
proto
  /vendor/
    <a>/
      Proto.toml
      ..
    <b>/
      Proto.toml
      ..
    <c>/
      Proto.toml
      ..
```

And simplifies code generation implementation.

## Upsides

This drastically simplifies the package store layout, and more importantly, enables people to compile it using *any* protobuf code gen tool, such as protoc without any change.

Furthermore it removes the type duplication problem that occured when nesting dependencies:

Before the change, you had two version of the `c` library types:

```rust
mod a {
  mod c { /* ... */ }
}

mod b {
  mod c { /* ... */ }
}
```

Which is now resolved, since tonic generates this structure:

```rust
mod a { /* ... */ }
mod b { /* ... */ }
mod c { /* ... */ }
```

## Downsides

This change has the implication that *all* dependencies may only use one single version of `c` which is not ideal. We can and will solve this by introducing proper version resolution mechanism soon.

### CODE REVIEW

File: src/generator.rs [hunk 1]
```diff
@@ -56,31 +45,23 @@ impl Generator {
...
-
-                let includes = &[&config.location];
-
                 tonic_build::configure()
                     .build_client(true)
                     .build_server(true)
                     .build_transport(true)
                     .compile_well_known_types(true)
-                    .out_dir(&out)
+                    .out_dir(output)
```
Comment (cdbrkfxrpt) [inline src/generator.rs:64]: if `output` is the output directory, and it is set via `.out_dir(...)`, i would probably simply name it `out_dir`

File: src/generator.rs [hunk 2]
```diff
@@ -117,33 +98,18 @@ pub async fn generate(language: Language) -> eyre::Result<()> {
         out
     };
 
+    generator
+        .run(&out)
```
Comment (cdbrkfxrpt) [inline src/generator.rs:102]: rename `out` to `out_dir` for clarity; avoids the reader having to read the code block to determine what `out` refers to

File: src/package.rs [hunk 3]
```diff
@@ -97,14 +95,29 @@ impl PackageStore {
...
         for (index, dependency) in dependencies.into_iter().enumerate() {
+            if let Ok(manifest) = Self::resolve(&dependency.package).await {
+                let existing = manifest.package.wrap_err(eyre::eyre!(
+                    "Found installed manifest for {} but it is malformed",
+                    dependency.package,
+                ))?;
+
+                eyre::ensure!(
+                    existing.version == dependency.manifest.version,
+                    "Another dependency in your project requires {}@{} which colides with {}@{} required by {}",
```
Comment (cdbrkfxrpt) [inline src/package.rs:109]:
```suggestion
                    "A dependency of your project requires {}@{} which collides with {}@{} required by {}",
```

**Review by cdbrkfxrpt — APPROVED** [review]
a few very minor nits aside, this lgtm

File: src/generator.rs [hunk 4]
```diff
@@ -56,31 +45,23 @@ impl Generator {
     pub const TONIC_INCLUDE_FILE: &str = "mod.rs";
 
     /// Run the generator for a dependency and output files into `out`
-    pub async fn run(&self, config: GenerationConfig<'_, '_>) -> eyre::Result<()> {
+    pub async fn run(&self, output: &Path) -> eyre::Result<()> {
```
Comment (wngr) [inline src/generator.rs:48]:
```suggestion
    pub async fn run(&self, output: impl AsRef<Path>) -> eyre::Result<()> {
```

File: src/generator.rs [hunk 5]
```diff
@@ -56,31 +45,23 @@ impl Generator {
     pub const TONIC_INCLUDE_FILE: &str = "mod.rs";
 
     /// Run the generator for a dependency and output files into `out`
```
Comment (wngr) [inline src/generator.rs:47]:
```suggestion
    /// Run the generator for a dependency and output files at the provided path
```

**Review by wngr — APPROVED** [review]

**Pull request merged**